### PR TITLE
feat(core): add MFA role to principal roles

### PR DESCRIPTION
**Pull request by Johaney-s**
* MFA role signals that principal is authorized by multifactor authentication
* this is resolved by obtaining info from UserInfoEndpoint
* if the acr value matches the one expected and timestamp is not older than allowed timeout, the role is added to principal
* the role itself will be required by critical operations in perun

### CODE REVIEW

**Review by cuadradek — APPROVED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java [hunk 1]
```diff
@@ -3946,4 +3957,57 @@ private static Map<String, Integer> createMappingOfValues(PerunBean complementar
...
+		String returnedTimestamp = updatedInformation.get(UserInfoEndpointCall.MFA_TIMESTAMP);
+		Instant parsedReturnedTimestamp;
+		try {
+			parsedReturnedTimestamp = Instant.parse(returnedTimestamp);
+		} catch (DateTimeParseException e) {
+			log.error("MFA timestamp "  + returnedTimestamp + " could not be parsed");
+			return false;
+		}
+		if (parsedReturnedTimestamp.isBefore(Instant.now())) {
+			log.error("MFA auth timestamp " + returnedTimestamp + " was grater than current time");
```
Comment (xflord) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java:4001]:
```suggestion
			log.error("MFA auth timestamp " + returnedTimestamp + " was greater than current time");
```
typo

**Review by xflord — APPROVED** [review]

**Review by xkureck — APPROVED** [review]

**Comment by Johaney-s** [comment]
changes after testing:
- renamed `userInfoEndpointAcrValue` to `userInfoEndpointMfaAcrValue`, because acr values are returned also by other authentication methods
- only MFA_TIMESTAMP is now added to additional informations if acr value matches mfa acr value, acr itself is not later needed
- it is now stored in principal's additional information so we can see updated value in `getPerunPrincipal` result, before it was only stored on creation and not updated there, separate map was used
- fixed wrong comparison of timestamp < now check
- moved retrieving issuer and access token in Api
- there's use case of delegation that is commented as fixme, we're okay keeping it out of MFA for now

**Review by cuadradek — APPROVED** [review]

**Review by balcirakpeter — CHANGES_REQUESTED** [review]
We do not need to explicitly call the `userInfoEndPoint` each time we refresh authorization. We need to retrieve information about MFA only during session refresh. Inside the authorization refresh, we work only with the given timestamp.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java [hunk 2]
```diff
@@ -3946,4 +3957,51 @@ private static Map<String, Integer> createMappingOfValues(PerunBean complementar
 
 		return mapping;
 	}
+
+	/**
+	 * Checks, if principal was authorized by Multi-factor authentication.
+	 * The information is resolved in UserInfoEndpointCall and stored in principal's additional informations.
```
[the remark this answers is not in the document]
Comment (Johaney-s) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java:3963]: it is consistent with the name of the property in perun principal

**Review by balcirakpeter — APPROVED** [review]

**Pull request merged**